File: src/me/xiaopan/android/imageloader/cache/disk/DiskCache.java
/*
 * Copyright (C) 2013 Peng fei Pan <sky@xiaopan.me>
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xiaopan.android.imageloader.cache.disk;

import me.xiaopan.android.imageloader.task.TaskRequest;

import java.io.File;

/**
 * 磁盘缓存器
 */
public interface DiskCache {
	/**
	 * 设置缓存目录
	 * @param cacheDir 缓存目录，在之后通过getCacheFile()方法获取缓存文件的时候就会在此文件夹下创建新文件
	 */
	public void setDir(File cacheDir);

    /**
     * 创建一个新的缓存文件
     * @param request 请求
     * @return 根据请求创建的缓存文件
     */
    public File createFile(TaskRequest request);

	/**
	 * 设置容量
	 * @param size 容量，在使用applyForSpace()方法申请空间的时候会计算当前缓存目录的容量是否超过此限制，如果超过就会按照活跃度（最后一次修改时间，每一次访问都会更新最后一次修改时间）来删除文件，直到腾出足够的空间
	 */
	public void setSize(long size);

	/**
	 * 申请空间
	 * @param length 尝试腾出足够的空间，删除的原则是按照活跃度（最后一次修改时间，每一次访问都会更新最后一次修改时间）来删除文件，直到腾出足够的空间
     * @return true：申请空间成功；false：申请空间失败
	 */
	public boolean applyForSpace(long length);

    /**
     * 清除缓存
     */
    public void clear();
}

File: sketch/src/main/java/com/github/panpf/sketch/cache/DiskCache.kt
/*
 * Copyright (C) 2019 panpf <panpfpanpf@outlook.com>
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.panpf.sketch.cache

import com.github.panpf.sketch.util.Logger
import kotlinx.coroutines.sync.Mutex
import java.io.Closeable
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream

/**
 * Disk cache for bitmap or uri data
 */
interface DiskCache : Closeable {

    companion object {
        const val DEFAULT_DIR_NAME = "sketch3"
    }

    var logger: Logger?

    /**
     * Get the cache directory on disk
     */
    val directory: File

    /**
     * Maximum allowed sum of the size of the all cache
     */
    val maxSize: Long

    /**
     * Sum of the size of the all cache
     */
    val size: Long

    /**
     * Returns an editor for the entry named [key], or null if another
     * edit is in progress.
     */
    fun edit(key: String): Editor?

    /**
     * Drops the entry for [key] if it exists and can be removed. Entries
     * actively being edited cannot be removed.
     *
     * @return true if an entry was removed.
     */
    fun remove(key: String): Boolean

    /**
     * Returns a snapshot of the entry named [key], or null if it doesn't exist.
     */
    operator fun get(key: String): Snapshot?

    /**
     * Returns exist of the entry named [key]
     */
    fun exist(key: String): Boolean

    /**
     * Clear all cached
     */
    fun clear()

    /**
     * Gets an edit lock bound to the specified [key], or creates a new one if it does not exist
     */
    fun editLock(key: String): Mutex

    /**
     * Snapshot the values for an entry.
     */
    interface Snapshot {
        /**
         * Returns cache key
         */
        val key: String

        /**
         * Returns cache file
         */
        val file: File

        /**
         * Returns the unbuffered stream
         */
        @Throws(IOException::class)
        fun newInputStream(): InputStream

        /**
         * Returns an editor, or null if another edit is in progress.
         */
        fun edit(): Editor?

        /**
         * Delete cache file
         *
         * @return If true is returned, the deletion is successful
         */
        fun remove(): Boolean
    }

    /**
     * Edits the values for an entry.
     */
    interface Editor {
        /**
         * Returns a new unbuffered output stream.
         * Call [commit] when you write done, or call [abort] if you don't want it
         */
        @Throws(IOException::class)
        fun newOutputStream(): OutputStream

        /**
         * Commits this edit so it is visible to readers.  This releases the
         * edit lock so another edit may be started on the same key.
         */
        @Throws(
            IOException::class,
        )
        fun commit()

        /**
         * Aborts this edit. This releases the edit lock so another edit may be
         * started on the same key.
         */
        fun abort()
    }
}